iwyu_globals.cc

```diff
@@ -177,6 +177,8 @@ OptionsParser::OptionsParser(int argc, char** argv) {
       intercepted_argv[intercepted_argc++] = argv[i];     // intercept --help
     else if (strcmp(argv[i], "--version") == 0)
       intercepted_argv[intercepted_argc++] = argv[i];     // intercept --version
+    else if (strcmp(argv[i], "-dM") == 0)
```

Comment:
While it's nice that this matches what you would do for the compiler, I wonder if it shouldn't be an IWYU option instead, i.e. something like `-Xiwyu --dump_macros`.

I had just started experimenting with getting IWYU to delegate preprocessing to Clang for `-dM -E`, but you beat me to it :)

Reply:
I think I have a plan, let me see if I can hack something up and post a branch.

Reply:
Ugh, I got stuck on weird API boundaries, and need to start thinking about dinner.

What I had in mind was roughly...

In iwyu_driver.cc
* only append `-fsyntax-only` if `-E` is NOT on the command-line
* search `compiler->getJobs()` for a job with action `PreprocessJobAction`; if it exists, go into preprocessing-only mode and don't run IWYU
* search `compiler->getJobs()` for a job with action `CompileJobAction`; if it exists, "replace" it by using `IwyuAction` as frontend action
* If neither exist, fail with "expected exactly one compiler job" or similar

Unfortunately I didn't find a nice way for iwyu_driver to communicate back to iwyu.cc's main which frontend action should be executed, etc.

I'll keep poking at it -- the goal is to allow `include-what-you-use -dM -E myfile.c` to follow clang's without us having to override their switches. And I think there may be other preprocessing tasks that would be useful if IWYU supported as normal.

Reply:
We could have IWYU `--dump_macros`... we just need to propagate `-dM` into `clang-args` to appropriately configure the preprocessing action. That would be an extra complication... hopefully not too bad.

Reply:
I think I've found the way:
https://github.com/kimgr/include-what-you-use/commit/detect-preprocess

This should make it possible to patch up the arguments, job list and driver options all in one place.

Reply:
Updated the branch with some semi-unrelated stuff; I've now smuggled the toolchain out into iwyu.cc so it can be used in `InitGlobals`, as I wanted to do earlier this year :)

Anyway -- based on this branch, I think I can "easily" let the driver decide what it wants to do, and then either replace compiler jobs with IWYU _or_ see that there's only preprocessing to be done, and let Clang handle that transparently.

Reply:
I kinda get what you're trying to do with the branch. Happy to rebase this over it when it's ready (dropping the first commit)

Reply:
Thanks! I'll try to finish it up, but I haven't had much focus time for IWYU lately.

Reply:
I think `include-what-you-use -dM -E` should work on master now